crates/video/src/stream.rs

```diff
@@ -0,0 +1,481 @@
+            inputq_resources: Default::default(),
+            outputq_resources: Default::default(),
+        })
+    }
+
+    pub fn memory(&self, queue_type: QueueType) -> MemoryType {
+        match queue_type {
+            QueueType::InputQueue => self.in_memory_type,
+            QueueType::OutputQueue => self.out_memory_type,
+        }
+    }
+
+    pub fn find_resource(&self, resource_id: u32, queue_type: QueueType) -> Option<&Resource> {
+        match queue_type {
+            QueueType::InputQueue => self.inputq_resources.map.get(&resource_id),
+            QueueType::OutputQueue => self.outputq_resources.map.get(&resource_id),
+        }
+    }
+
+    pub fn find_resource_mut_by_index(
+        &mut self,
+        v4l2_index: u32,
+        queue_type: QueueType,
+    ) -> Option<&mut Resource> {
+        match queue_type {
+            QueueType::InputQueue => self.inputq_resources.find_mut_by_index(v4l2_index),
+            QueueType::OutputQueue => self.outputq_resources.find_mut_by_index(v4l2_index),
+        }
+    }
+
+    pub fn find_resource_mut(
+        &mut self,
+        resource_id: u32,
+        queue_type: QueueType,
+    ) -> Option<&mut Resource> {
+        match queue_type {
+            QueueType::InputQueue => self.inputq_resources.map.get_mut(&resource_id),
+            QueueType::OutputQueue => self.outputq_resources.map.get_mut(&resource_id),
+        }
+    }
+
+    pub fn is_queue_streaming(&self, queue_type: QueueType) -> bool {
+        match queue_type {
+            QueueType::InputQueue => self.inputq_resources.streaming,
+            QueueType::OutputQueue => self.outputq_resources.streaming,
+        }
+    }
+
+    pub fn set_queue_streaming(&mut self, queue_type: QueueType) {
+        match queue_type {
+            QueueType::InputQueue => self.inputq_resources.streaming = true,
+            QueueType::OutputQueue => self.outputq_resources.streaming = true,
+        }
+    }
+
+    pub fn set_state(&mut self, state: StreamState) {
+        self.state.set_state(state);
+    }
+
+    pub fn state(&self) -> StreamState {
+        self.state.state()
+    }
+
+    pub fn all_created(&self, queue_type: QueueType) -> bool {
+        self.all_resources_state(queue_type, ResourceState::Created)
+    }
+
+    #[allow(dead_code)]
+    pub fn all_queued(&self, queue_type: QueueType) -> bool {
+        self.all_resources_state(queue_type, ResourceState::Queued)
+    }
+
+    fn all_resources_state(&self, queue_type: QueueType, state: ResourceState) -> bool {
+        match queue_type {
+            QueueType::InputQueue => self
+                .inputq_resources
+                .resources()
+                .into_iter()
+                .all(|x| x.state() == state),
+            QueueType::OutputQueue => self
+                .outputq_resources
+                .resources()
+                .into_iter()
+                .all(|x| x.state() == state),
+        }
+    }
+
+    #[allow(dead_code)]
```

Comment:
If this is not needed right now, can it be added later?

Reply:
Yes, no problem.